File: backend/app/services/gemini.py
import base64
import mimetypes

from google import genai

from app.config import GEMINI_API_KEY

client = genai.Client(api_key=GEMINI_API_KEY) if GEMINI_API_KEY else None


async def transcribe_audio(audio_data: bytes, filename: str, language: str) -> str:
    if not client:
        return ""

    mime_type, _ = mimetypes.guess_type(filename)
    if not mime_type:
        mime_type = "audio/wav"

    prompt = (
        f"Transcribe the speech in this audio file to plain text. "
        f"The spoken language is {language}. "
        f"Return only the transcribed text, no explanations."
    )

    try:
        resp = client.models.generate_content(
            model="gemini-2.0-flash",
            contents=[prompt, genai.types.Part.from_bytes(data=audio_data, mime_type=mime_type)],
        )
        return resp.text.strip() if resp.text else ""
    except Exception:
        return ""


async def translate_text(text: str, source: str, target: str) -> str:
    if not client:
        return ""

    prompt = (
        f"Translate the following text from {source} to {target}. "
        f"Return only the translated text, no explanations.\n\n{text}"
    )

    try:
        resp = client.models.generate_content(model="gemini-2.0-flash", contents=[prompt])
        return resp.text.strip() if resp.text else ""
    except Exception:
        return ""

File: backend/src/services/gemini.js
import { GoogleGenAI, createPartFromBase64 } from "@google/genai";
import { GEMINI_API_KEY } from "../config.js";

const client = GEMINI_API_KEY ? new GoogleGenAI({ apiKey: GEMINI_API_KEY }) : null;
const MODEL = "gemini-2.0-flash";

const LANGUAGE_NAMES = { en: "English", ja: "Japanese" };

// A 429 here means the API key's quota is exhausted, not a brief rate spike
// (Google reports "limit: 0" until billing/quota is fixed) — every retry
// wastes a full round trip before falling back. Skip Gemini for a cooldown
// window after a quota error instead of eating that latency on every call.
const QUOTA_COOLDOWN_MS = 60_000;
let quotaCooldownUntil = 0;

function isQuotaError(err) {
  return err?.status === 429;
}

export async function transcribeAudio(audioBuffer, filename, language) {
  if (!client || Date.now() < quotaCooldownUntil) return { text: "", language: null };

  const mimeType = filename?.endsWith(".mp3")
    ? "audio/mp3"
    : filename?.endsWith(".ogg")
      ? "audio/ogg"
      : filename?.endsWith(".m4a")
        ? "audio/mp4"
        : "audio/wav";

  const isBilingual = language === "en-ja";
  const prompt = isBilingual
    ? 'Listen to this audio and determine whether the speaker is speaking English or Japanese. ' +
      'Transcribe exactly what was said, verbatim, in that language\'s native script (do not translate ' +
      'or romanize Japanese speech). Respond with strict JSON only, no markdown fences: ' +
      '{"language": "en" or "ja", "text": "<transcript>"}'
    : `Transcribe the speech in this audio file to plain text. The spoken language is ${LANGUAGE_NAMES[language] || language}. Return only the transcribed text, no explanations.`;

  try {
    const resp = await client.models.generateContent({
      model: MODEL,
      contents: [
        prompt,
        createPartFromBase64(audioBuffer.toString("base64"), mimeType),
      ],
    });

    const raw = resp?.text?.trim() || "";
    if (!isBilingual) return { text: raw, language: null };

    try {
      const cleaned = raw.replace(/^```json\s*|\s*```$/g, "");
      const parsed = JSON.parse(cleaned);
      return {
        text: typeof parsed.text === "string" ? parsed.text : "",
        language: parsed.language === "ja" || parsed.language === "en" ? parsed.language : null,
      };
    } catch {
      return { text: raw, language: null };
    }
  } catch (err) {
    if (isQuotaError(err)) quotaCooldownUntil = Date.now() + QUOTA_COOLDOWN_MS;
    return { text: "", language: null };
  }
}

export async function translateText(text, source, target) {
  if (!client || Date.now() < quotaCooldownUntil) return "";

  try {
    const resp = await client.models.generateContent({
      model: MODEL,
      contents: [
        `Translate the following text from ${source} to ${target}. Return only the translated text, no explanations.\n\n${text}`,
      ],
    });

    return resp?.text?.trim() || "";
  } catch (err) {
    if (isQuotaError(err)) quotaCooldownUntil = Date.now() + QUOTA_COOLDOWN_MS;
    return "";
  }
}
